### ISSUE TITLE
DCOS Sub Command CLI has problems with POSTing multipart

### ISSUE THREAD

**Issue by huntc** [dcos/dcos-cli#818]
Please answer the following questions before submitting your issue. Thanks!
### What version of DC/OS + DC/OS CLI are you using (`dcos --version`)?

dcoscli.version=0.4.10
dcos.version=1.8.4
dcos.commit=e64024af95b62c632c90b9063ed06296fcf38ea5
dcos.bootstrap-id=5b4aa43610c57ee1d60b4aa0751a1fb75824c083
### What operating system and version are you using?

OS X 10.11.6
### What did you do?

I'm writing a custom sub command and I'm using the DC/OS Python http library (0.4.14). When performing a POST request that uploads files I receive a bad request status back. The reason for this is because the DC/OS implementation of the post request makes several attempts. The first attempt is without authorisation, and the subsequent ones are with authorisation. Because I'm uploading a multipart file, and the stream associated with the request isn't reset, no content will be sent on the subsequent requests. Here's how I prepare the request:

```
        files.append(('bundle', (bundle_name, open(bundle_file, 'rb'))))
```
### What did you expect to see?

That my load request succeed even on subsequent attempts.
### What did you see instead?

The first load correctly streams the parts, but the subsequent ones do not. The first one fails though because authorisation is required.
---

The solution is for your flavour of the requests post method to scan through the `files` data structure and reset any stream found by calling `seek(0)`.

Even better, why even retry? Why not just one request with authorisation always? I'd state that you've broken the contract of the request post method, and that perhaps this should be some higher level function.

**Comment by tamarrow** [comment]
We retry if we get a 401 because we use the `www-authenticate` header to find out which type of auth the server is looking for (if any). You can always check that a user is authorized before this command is run, and then it will only run one request. Or potentially read the file, and send it in as a payload- assuming the file is small. But, we don't want to globally call `seek(0)` because it will break streaming requests. 

The requests library does not support streaming for multi-part files: http://docs.python-requests.org/en/master/user/quickstart/#post-a-multipart-encoded-file
How are you sending the request?

**Comment by huntc** [comment]
Can you not use a `HEAD` request to first determine the auth server i.e. it doesn't have to be the actual request right?

The requests library accepts a stream from the `open` built-in function that you cite. That's the type of stream I'm referring to.

So now just two questions:

1. Why not use `HEAD` in order to determine the auth server?
2. Why perform a request retry at all? 

On (2), this violates the existing "contract" of the requests library IMHO, and serves no useful purpose so far as I can see.

**Comment by tamarrow** [comment]
You are right, we could use HEAD to determine the auth server and that would fix your problem. Currently we assume no auth, unless we see it. Your approach would be the opposite, but would work too.

**Comment by huntc** [comment]
Would you be open to a PR from me that uses HEAD and also removes the retry logic that you have?

**Comment by tamarrow** [comment]
Adding a PR to switch to using HEAD would be great! Lets keep the password retry code though.

**Comment by huntc** [comment]
Could you please explain why is the password retry code is required?

The problem here is that you can't retry the request given the scenario where the request contains stream objects that have been consumed previously.

For me, it feels like the retry logic should be some higher level concern, and not a part of the request. WDYT?

**Comment by tamarrow** [comment]
It's common that a user gets to try their password a few times (typing is hard). We can retry the HEAD request multiple teams to get valid login creds, and only try the actual request once.

**Comment by huntc** [comment]
Ah got it - you're relying on interactive password prompting...

Ok, so multiple HEAD requests, single actual request. I'll see if I can get you a PR today.

**Comment by tamarrow** [comment]
👍

**Comment by tamarrow** [comment]
Fixed with #824
